### PR TITLE
[codegen] Improved handling of records and record arrays in the codegen.

### PR DESCRIPTION
**Pull request by mahge**
- Every record now gets
  - A dedicated "default" constructor function for allocating its members.
    - this is used when the variable is declared. It initializes members to
      their default values.
    - Note that this is different from default Modelica record constructor.
      The modelica constructor is used explicitly. The "default" one is used
      always on declaration.
  - A dedicated copy function for copying its members.
  - An array typedef to base_array.
  - A macro for creating an array of the record.
    - This uses the constructor function to create each element of the array.
    - This uses generic array function with the record size.
  - A macro for copying an array
    - This uses the copy function to copy each element of the array.
  - A macro getter to access elements of the array.
    - this uses the proper casting on the returned type.

- We now keep bindings in TYPES_VARS in when creating a record declaration
  in SimCode structure. The bindings are used as default values by the "default"
  constructor.

- Bindings of a derived record declaration are handled properly now.
  - If a record is a derived record with modifications, then bindings of the modifications
    are available in the declration scope.
         record A = B(k=exp); A r; //  'exp' needs to be passed to the constructor
      => A_construct(A* r, exp) {...; r->k = exp; ...}
    This means a default constructor will need those bindings passed to it explicitly
    if it is to work correctly. Now every TYPES_VAR of a record has a Boolean attribute
    specifying whether it is bound from outside or not.

- When collecting used functions in the BACK-END we traverse bindings of record members.
  - Right now this is needed because some functions get in-lined and are not used by the
    equation system even though the model has instances of the record (since we generate
    constructors now we need them to be visible even if the record is never used in a function).
  - The proper way to fix this would be to check if a record is used in a function and
    only then generate a constructor. OR another option is to also traverse record member
    bindings when doing in-lining of functions (which is not done right now.)
    This seemed to be the simplest and probably quickest way even though we sometimes may
    generate a function even though it is not really used (very rarely).

- Fixed copying of records in the simulation context.

  - If a record is being copied to simulation vars we need special handling.
    This is because there is no structure to simvars. The variables are scattered.
    Therefore there is now a copy function that writes a given record to the corresponding
    element in the simvars arrays.

    The generated function <RecName>_copy_to_vars(RecType* src, .../*simvars locations*/) is used to achive this.

- Improved generation of component references in function context
  - We can now handle qualified crefs with subscripts anywhere in the cref.
  - This is done uniformly in functionContextCref template. Try to use this everywhere.

- CONTEXT_FUNCTION now has a cref prefix
  - This is prepended to any cref generated in function context. This gives us some scope
    control. E.g. it is used in record constructor functions to make all local variable accesses
    prepended to a specific record name that is passed as input to the function.

- instDims in SimCodeFunction.VARIABLE is now list<DAE.dimension> not  list<DAE.Expression>.
  - Helps to normalize handling of dimension expressions.
  - We really need to normalize the handling of dimension and subscript
  - Plus dimension and subscript related functions are not exchangeable.
    Constructs are interpreted differently as dimension compared to subscript
    and vice versa.

- Unknown size array handling.

  - Arrays of unknown size are arrays where at least one dimension is unknown.
  - If an unknown size array is declared with a default value, i.e. a binding,
    then it has fixed dimension sizes equal to the binding.
  - However if the unknown size array is declared with out a binding then it
    is considered flexible and can change sizes as needed.
  - The array representation in the runtime system ,i.e. base_array, now has a
    new memeber .flexible to signify this.

- Handle array expressions of records i.e. {c1,c2,..} properly
  - Array expressions of records are handled a bit more cleanly now.
  - Do not sort record member variables in the record declaration. This is just absurd.

- Some more minor fixes.
  - Casting of call expressions which return records is handled explicitly now.
  - Change handling of casts.
    - Casts to Modelica_integer are disabled for now.
    - Casts to records with different number of members are disabled for now.
    - Many more...

### CODE REVIEW

File: OMCompiler/Compiler/Util/List.mo [hunk 1]
```diff
@@ -1011,26 +1011,6 @@ algorithm
   outList := append_reverse(outList, l1);
 end mergeSorted;
 
-public function sortIntN
```
Comment (kabdelhak) [inline OMCompiler/Compiler/Util/List.mo]: Why did you remove this function? Is it not used anymore? Seems reasonable to have a linear sorting function if the maximum value is known.

File: OMCompiler/SimulationRuntime/c/util/integer_array.c [hunk 2]
```diff
@@ -93,17 +94,7 @@ void alloc_integer_array_data(integer_array_t* a)
 
 void copy_integer_array_data(const integer_array_t source, integer_array_t* dest)
 {
```
Comment (kabdelhak) [inline OMCompiler/SimulationRuntime/c/util/integer_array.c]: Why these wrapper functions? copy_integer_array_data is just calling integer_array_copy_data...

File: OMCompiler/Compiler/Util/List.mo (hunk 1 above)
Comment (mahge) [reply OMCompiler/Compiler/Util/List.mo]: 1) The complexity of this function is not linear on the input size (which is what I think you are implying) it is O(inN). In other words, say if you want to sort {2,1,3,1000} this 1000x in ops and memory (compared to say O(n*log(n)) = 8x). Not to mention the other non-asymptotic things like allocating such a big new array and initializing it.

2) let's say this linear. How often do you sort large lists of 'n' integers between 1 and N where n~N?

3) Do you need this function right now? Do you think it can be implemented when and if needed?

4) Is this really an important issue in this PR? Are there no other things like overall design and implementation choices that you want to address after going through it twice?

Comment (sjoelund) [reply OMCompiler/Compiler/Util/List.mo]: Well, it's O(max (N,inN)). For inN = 1000 and sorting around 300 elements you break even. It should probably be called radix sort if it remained. One of the benefits of it is that no comparison is performed. If using a loop it should probably pipeline rather well especially since this version does not count number of hits.

Comment (JKRT) [reply OMCompiler/Compiler/Util/List.mo]: To be fair I discussed this with Mahder, and supported it's removal. The reasoning was that it had wrong naming, did not handle duplicates and the use of it was removed if I remember. I am also to be blamed!

Comment (mahge) [reply OMCompiler/Compiler/Util/List.mo]: > Well, it's O(max (N,inN)). 
If n > inN the input list is not unique. In that case the function does sort + unique in O(n) time and gives back inN elements. While that is still a valid operation, I think it is not what the function was intended to do.

Comment (kabdelhak) [reply OMCompiler/Compiler/Util/List.mo]: As an answer to 4.: Everything else seems fine, these were the only two small things i recognized.

**Comment by adrpo** [comment]
@mahge can you push a change to this so is ran again via the CI?

**Comment by adrpo** [comment]
I added labels to check also the build on MinGW and MacOS.

**Comment by adrpo** [comment]
ok, let's see how it goes!

**Comment by adrpo** [comment]
https://test.openmodelica.org/jenkins/job/LibraryTesting/job/LibraryTest/830/console
I guess this one is including the changes.

**Comment by adrpo** [comment]
Doesn't sound so good:
newInst 2019-10-02 20:28:55..2019-10-03 11:29:46.html 5 improved, 342 regressions; performance 0 improved, 1 regressions
We'll see how it looks like when the report actually finishes generating :)

**Comment by adrpo** [comment]
https://libraries.openmodelica.org/branches/history/newInst/2019-10-02%2020:28:55..2019-10-03%2011:29:46.html

**Comment by adrpo** [comment]
Seems to be the same issue: 
```
Buildings_latest_Buildings.Fluid.Examples.Performance.Example1v1_records.c:60:44: error: no member named '_per' in 'Buildings_Fluid_Movers_BaseClasses_Characteristics_efficiencyParameters'
  tmp0 = size_of_dimension_base_array(ths->_per._V_flow, ((modelica_integer) 1));
                                      ~~~  ^
```
Hopefully you can fix it easily.

**Comment by mahge** [comment]
I think I know what this caused this. The derived record binding expressions issue that I fixed for the old front-end. I just set the "marker" thing to false when NF converts to DAE and then forgot about it. I will fix it. Thanks for the heads up.

I am worried about the other ones though :) We will see. I am going through the results.

**Comment by mahge** [comment]
https://libraries.openmodelica.org/branches/history/master/2019-10-02%2018:39:58..2019-10-03%2002:14:49.html

64 regressions on master. I will check them and also try to add smaller tests for the issues.
